Initial implementation

corvus-ch opened the pull request.

Commit:
Define default variables

Commit:
Meet preconditions and create LV

Commit:
Create file system

Commit:
Create mount

Commit:
Ensure file permissions

Commit:
Do nothing for absent volumes

Commit:
Mount by UUID

Commit:
Handle unmount of removed volumes

Commit:
Handle volume removal

hansmi commented on defaults/main.yaml:
```diff
@@ -0,0 +1,18 @@
+  - lvm2
+  - xfsprogs
+capra_lvm_filesystem_lvcreate_options: "--wipesignatures y --zero y"
+capra_lvm_filesystem_filesystem_options: "-i size=512 -n size=8192"
```
These _may_ be the default values built into `mkfs.xfs`. Can you verify?

corvus-ch replied:
The are different. For `-I` the man page says:
> […] The minimum (and default) value is 256 bytes […]
And for `-n` the man page says:
> […] The default size value for version  2  directories  is 4096  bytes  (4  KiB),  unless the filesystem block size is larger than 4096, in which case the default value is the filesystem  block size […]

But the actual question is whether we should have them here as they are xfs specific and the role can be used for any file system.

hansmi replied:
One could argue that the role needs to define a default filesystem and may as well define filesystem options while retaining the ability to override both. In the other case the role shouldn't install `xfsprogs` either.

corvus-ch replied:
Makes sense. As the options deviate from the defaults, let us keep them.

hansmi commented on meta/main.yml:
```diff
@@ -0,0 +1,6 @@
+  description: Ansible role for managing logical volumes on LVM, ensure they have a file system and are mounted at some place.
+  company: VSHN AG
+  license: BSD
+  min_ansible_version: 2.6
```
Needs to be supported with Ansible 2.4.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,77 @@
+  package:
+    name: "{{ item }}"
+    state: present
+  with_items: "{{ capra_lvm_filesystem_packages }}"
```
`package` accepts a list directly.

corvus-ch replied:
Where is this documented?

hansmi replied:
I figured out by reading code and lots of our other Ansible roles make use of that feature.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,77 @@
+  mount:
+    path: "{{ item.mount | default(capra_lvm_filesystem_mount_parent + '/' + item.name) }}"
+    state: absent
+  with_items: "{{ capra_lvm_filesystem_specs }}"
```
It's a list, so `with_list` (applies to all other cases too).

corvus-ch replied:
In order to remain compatible with Ansible 2.4 we have to use `with_items`. See https://docs.ansible.com/ansible/2.4/playbooks_loops.html.

hansmi replied:
`with_list` is supported in 2.4 and widely used.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,77 @@
+
+- name: "Unmount volume"
+  mount:
+    path: "{{ item.mount | default(capra_lvm_filesystem_mount_parent + '/' + item.name) }}"
```
Use `~` to concatenate values. `mount` can mean many things, hence other places use `mountpath`. Also, why duplicate this logic multiple times instead of defining a module-global variable making use of `item`? Now that it's mentioned: Rename the loop variable with a prefix.

corvus-ch replied:
There is still one instance of duplication left in the task using `with_indexed_items`. Any suggestions to get rid of that one?

corvus-ch replied:
Found a solution for that one.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,77 @@
+    size: "{{ item.size | human_to_bytes }}B"
+    opts: "{{ capra_lvm_filesystem_lvcreate_options }}"
+    state: "{{ item.state | default('present') }}"
+    force: "{{ (item.state | default('present')) == 'absent' }}"
```
I think it'd be better to split out the `present` and `absent` paths into separate, imported files. That way tags can also be used to selectively skip operations.

corvus-ch replied:
I kept it all in one file as there are less than 100 lines but I splitter the paths and added tags.

hansmi replied:
Responded on later revision.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,77 @@
+
+- name: "Set volume permissions"
+  file:
+    path: "{{ item.mount | default(capra_lvm_filesystem_mount_parent + '/' + item.name) }}"
```
See above, code duplication.

Commit:
Deduplicate variable defaults

At least most of them.

tobru commented on README.md:
```diff
@@ -58,24 +58,36 @@ and are mounted.
   roles:
     - role: "capra-lvm-filesystems"
-      capra_lvm_filesystem_mount: "/var/lib/rancher/ssd"
+      capra_lvm_filesystem_mount_parent: "{{ vshn_sbar_local_volume_path }}/ssd"
```
remove vshn and sbar from example, should be more anonymous

tobru commented on README.md:
```diff
@@ -58,24 +58,36 @@ and are mounted.
 	  size: 5G
     - role: "capra-lvm-filesystems"
-      capra_lvm_filesystem_mount: "/var/lib/rancher/ssd-retain"
+      capra_lvm_filesystem_mount_parent: "{{ vshn_sbar_local_volume_path }}/ssd-retain"
```
remove vshn and sbar from example, should be more anonymous

tobru commented on README.md:
```diff
@@ -58,24 +58,36 @@ and are mounted.
+      capra_lvm_filesystem_mount_parent: "{{ vshn_sbar_local_volume_path }}/ssd"
       capra_lvm_filesystem_specs:
         - name: "{{ inventory_hostname + '7jOa' | to_uuid }}"
 	  size: 5G
```
Do not use tabs, use plain spaces

tobru reviewed (dismissed):
The README needs some cleanup:
* Variables listing is broken in rendered file
* Replace tabs by spaces

hansmi commented on README.md:
```diff
@@ -58,27 +58,39 @@ and are mounted.
                                 Defaults to `defaults`.
+## Tags
+
+The following tags can be used to controll the executed actions:
```
`s/controll/control/`

hansmi commented on README.md:
```diff
@@ -37,11 +37,11 @@ and are mounted.
-				     Defaults to `/mnt`.
+                               Defaults to `xfs`.
+`capra_lvm_filesystem_parent_path`: Default path where the volumes will be
+                                    mounted to.  The name of the volume will
```
Not sure how the format of this list is supposed to work. It clearly doesn't. Use a plain itemized list:

```
* `foo:` bar with long lines
  when wrapped they
  are not indented
* `bar`: blubb
```

hansmi commented on README.md:
```diff
@@ -58,27 +58,39 @@ and are mounted.
                                 Defaults to `s0`.
 `capra_lvm_filesystem_options`: Default mount point options.
```
String or list?

hansmi commented on README.md:
```diff
@@ -58,27 +58,39 @@ and are mounted.
+
+The following tags can be used to controll the executed actions:
+
+* `capra_lvm_filesystem_prerequisites`:
```
Tags use dashes in our other code.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+  tags: [capra_lvm_filesystem_prerequisites]
+
+- vars:
+    _capra_lvm_filesystem_path: "{{ item.path | default(capra_lvm_filesystem_parent_path ~ '/' ~ item.name) }}"
```
"Local" variables in our other code use `r_` as a prefix (from `register`, even though they're not always registered variables).

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+    _capra_lvm_filesystem_path: "{{ item.path | default(capra_lvm_filesystem_parent_path ~ '/' ~ item.name) }}"
+    _capra_lvm_filesystem_state: "{{ item.state | default('present')}}"
+  block:
+    - name: "Unmount volume"
```
No quotes necessary in such cases

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+---
+- name: Install required pakages
```
`s/pakages/packages/`

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+  package:
+    name: "{{ item }}"
+    state: present
+  with_list: "{{ capra_lvm_filesystem_packages }}"
```
YUM on CentOS/RHEL can be rather slow. Please test whether the impact of invoking the module twice instead of once (by passing the list directly) isn't significant.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+---
+- name: Install required pakages
+  package:
+    name: "{{ item }}"
```
Set a custom loop variable name.

corvus-ch replied:
`item` is a common thing in Ansible. While it is possible to rename it pre 2.7, the output might be buggy. Therefore I will leave it like that.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+        path: "{{ _capra_lvm_filesystem_path }}"
+        state: absent
+      with_list: "{{ capra_lvm_filesystem_specs }}"
+      when: _capra_lvm_filesystem_state == 'absent'
```
I _really_ dislike mixing two distinct code paths. If you don't want to use `include_tasks` (understandable) you can probably get away with this (still results in static imports):
```
- when: "_capra_lvm_filesystem_state == 'absent'"
  import_tasks: volume-absent.yml

- when: "_capra_lvm_filesystem_state == 'present'"
  import_tasks: volume-present.yml
```

corvus-ch replied:
It actually added a whole bunch of issues I had to solve, but I did it none the less.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+      when: _capra_lvm_filesystem_state == 'present'
+      tags: [capra_lvm_filesystem_create]
+
+    - name: "Lookup volume UUIDs"
```
"Lookup" is different from "Look up". The latter is correct here.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+      lvol:
+        vg: "{{ item.vg | default(capra_lvm_filesystem_vg) }}"
+        lv: "{{ item.name | mandatory }}"
+        size: "{{ item.size | human_to_bytes }}B"
```
No need for the size and options when removing.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+      tags: [capra_lvm_filesystem_create]
+
+    - name: "Lookup volume UUIDs"
+      command: "/sbin/blkid -c /dev/null -o value -s UUID {{ '/dev/' ~ ((item.vg | default(capra_lvm_filesystem_vg)) ~ '/' ~ item.name) | quote }}"
```
Wrap to multiple lines <80 characters. `>-` is your friend.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+    - name: "Set volume permissions"
+      file:
+        path: "{{ _capra_lvm_filesystem_path }}"
+        state: directory
```
This should probably not create the directory if it weren't to exist for some reason. That must happen before the mount

corvus-ch replied:
Got that covered by not touching the files if the mount failed.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,100 @@
+        fstype: "{{ item.1.fstype | default(capra_lvm_filesystem_fstype) }}"
+        opts: "{{ item.1.options | default(capra_lvm_filesystem_options) }}"
+        state: mounted
+      with_indexed_items: "{{ capra_lvm_filesystem_specs }}"
```
Instead of iterating over the volumes you could iterate over the results of the UUID lookup. Every result item contains its input item (use the `debug` module to observe the structure). That's a technique I've successfully used in other places.

corvus-ch replied:
It kinda does so now.

Commit:
Do not fail resize if volume is not mounted

The `filesystem` fails to resize if the fileystem is not mounted. In
order to not fail a resize, resizing was moved after the call to the
`mount` module. The original call to `filesystem` only takes care of
creating the filesystem but ignores changes to its size.

Commit:
Unified variable naming in loops

Have consistent naming of variables for different types of loops. This
reduces copies of similar logic but different variable names.

Commit:
Do not touch permissions expept after creation

Programms running on a system might take controll of file permissions.
In order to prevent games of ping pong, the permissions of new volumes
are only touchged right after a volume gets created. Setting permissions
is required to have sensible and known defaults.

Commit:
Prevent creation of empty directory

When creating a new volume, if the mount fails for some reason, an
empty directory would be created while setting the volumes initial
permissions. Preventing this by only set permissions if the mount was
successful.

hansmi commented on tasks/main.yaml:
```diff
@@ -0,0 +1,31 @@
+
+- vars:
+    r_capra_lvm_filesystem_path: "{{ item[1].path | default(capra_lvm_filesystem_parent_path ~ '/' ~ item[1].name) }}"
+    r_capra_lvm_filesystem_state: "{{ item[1].state | default('present')}}"
```
I think you could've kept this code simpler by using [Jinja's `groupby` filter](http://jinja.pocoo.org/docs/2.9/templates/#groupby) to prepare two separate lists which are then processed separately. Is this something you'd like to address before merging?

hansmi commented on tasks/volume-present.yaml:
```diff
@@ -0,0 +1,69 @@
+  register: r_capra_lvm_filesystem_mount
+  with_together:
+    - "{{ r_capra_lvm_filesystem_blkid.results }}"
+    - "{{ capra_lvm_filesystem_specs }}"
```
`r_capra_lvm_filesystem_blkid.results` should already contain the specs as part of the input item (use the `debug` module and you should see the structure).
